snyk/snyk
# Expose snyk/snyk-iac-parsers via gopherjs [cfg-1499]

#### What does this PR do?
In https://github.com/snyk/snyk-iac-parsers, we are building a new functionality, where we will parse HCL files and dereference variables.

We will want to integrate this with the Snyk CLI, and as a first step we need to expose the function of the specific Go package. 

This PR integrates and exposes the needed function from https://github.com/snyk/snyk-iac-parsers.

We chose GopherJS for the compilation of Go->JS as this has been used before from our team for similar purposes. The readme, Makefile and general idea is a duplicate of the existing https://github.com/snyk/snyk/tree/master/release-scripts/hcl-to-json-parser-generator.  This will generate two files a hcltojson-v2.js and an hcltojson-v2.js.map with
source maps for the minified file. The output file is ~4.0M minified and ~716K gzipped.
You can find these files under `/release-scripts/hcl-to-json-parser-generator-v2/src/hcltojson-v2/dist/`

For now, we have suffixed this new version as `v2`. 
The plan is to remove the old `hcl-to-json` directory when when the new functionality is released.
 
#### Where should the reviewer start?

You can just call the function like this:  
```
parseModule({
    '../path/to/terraform.tf': fileContent,
    'file2.tf': 'content2',
  });
```

Following the contract established on https://github.com/snyk/snyk-iac-parsers/pull/6, the parser expects a map of files in the form of (input):

```
{
   'open_ssh.tf': `resource "aws_security_group" "allow_ssh" {
  name        = "allow_ssh"
  description = "Allow SSH inbound from anywhere"
  cidr_blocks = var.dummy
}
`,
   's3_bucket_acl.tf': `resource "aws_s3_bucket" "s3_bucket" {
  bucket = var.bucket_name

  acl    = "public-read"
  website {
    index_document = "index.html"
    error_document = "error.html"
  }

  tags = var.tags
}
`,
}
```
and it returns an output of :

```
{
parsedFiles: {},
failedFiles: {}
```

Co-authored-by: Teodora Sandu <teodora.sandu@snyk.io>

#### How should this be manually tested?

Run `make build && make test` locally to see the assertions in `test.js` succeed. 


#### Any background context you want to provide?

This is the first step of supporting Variables in Terraform scanning. There is going to be another PR calling this exposed function for iac test and will be behind a feature flag in the beginning.

For even wider context, the first time we introduced a JS bundle via GopherJS: https://github.com/snyk/snyk/pull/1849

#### What are the relevant tickets?
[[CFG-1499](https://snyksec.atlassian.net/browse/CFG-1499)]
